ls1intum/Artemis
# Bugfix/2752 bulk editing removes categories

### Checklist

### Motivation and Context


Fixes #2752 where categories of Programming Exercises are removed when 'bulk updated'.

In addition, the optional 'Notification Text' has been added to the 'Edit selected' modal, which was missing before.

### Description

A new Endpoint **PUT /programming-exercises/timeline** has been implemented to only update the timeline of a Programming Exercise. This fixes the before mentioned issue and improves the performance of the bulk updates.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Select one or more Programming Exercises containing at least one 'Category'
4. Click on 'Edit Selected'
5. Choose a Release and Due date
6. Click 'Save'
7. The selected Exercises should now have the new release and due dates and their 'categories' should have remain untouched.

### Test Coverage



ProgrammingExerciseResourc.javae 82%
ProgrammingExerciseService.java 92%

programming-exercise-edit-selected.component.ts 87.27%
programming-exercise.service.ts 65.15%


### Screenshots


The *Edit Selected* modal with optional 'notification text field'
![Screenshot from 2021-01-28 17-16-51](https://user-images.githubusercontent.com/45761921/106166823-db731480-618c-11eb-871b-ed79424c1200.png)
